### PR TITLE
impl(bigquery): `DataClient` and `*Connection`

### PR DESCRIPTION
**Pull request by jsrinnn**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11414)

### CODE REVIEW

File: google/cloud/bigquery/v2/minimal/internal/dataset_client.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  friend bool operator==(DatasetClient const& a, DatasetClient const& b) {
+    return a.connection_ == b.connection_;
+  }
+  friend bool operator!=(DatasetClient const& a, DatasetClient const& b) {
+    return !(a == b);
+  }
+
+  /// Gets specific dataset information from Bigquery specific to the datasetID
+  /// specified in the request. For more details on BigQuery datasets, please
+  /// refer to: https://cloud.google.com/bigquery/docs/managing-datasets
```
Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/dataset_client.h:56]: Optional: if we are going to use Doxygen comments I would have said:

```
/// Gets the metadata for the given dataset.
///
/// @see https://cloud.google.com/bigquery/docs/managing-datasets for more details on BigQuery datasets.
```

Strangely we tend to use `/**` blocks in hand-crafted code, but the generator uses `///` blocks.  They mean the same thing.

Comment (coryan) [reply google/cloud/bigquery/v2/minimal/internal/dataset_client.h:56]: BTW, by convention the first paragraph in a Doxygen comment is the 'brief' description.  So it tends to be short.

File: google/cloud/bigquery/v2/minimal/internal/dataset_rest_connection_impl.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                            Options options);
+
+  Options options() override { return options_; }
+
+  StatusOr<Dataset> GetDataset(GetDatasetRequest const& request) override;
+  StreamRange<ListFormatDataset> ListDatasets(
+      ListDatasetsRequest const& request) override;
+
+ private:
+  std::unique_ptr<DatasetRetryPolicy> retry_policy() {
```
Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/dataset_rest_connection_impl.h:49]: I think @dbolduc had opinions as to whether these helpers actually help.

**Review by coryan — APPROVED** [review]
A small nit:

By convention, commit titles should be < 72 characters, and some people aim for 50.  I know this feels restricting.  I would have said:

```
impl(bigquery): `DataClient` and `*Connection`
````

Note that conventional commits recommend no period at the end:

https://www.conventionalcommits.org/en/v1.0.0/#specification

I think the class documentation would not be enough for a public API, but this is not a public API.

**Review by jsrinnn — COMMENTED** [review]
Ack. Thanks for the link. 

Modified the commit title for the current PR based on above. Addressed other comments as well.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11414#-:-NU2u7THEQGaLwr3UbFo:bsvo7wz)* status: 0 of 13 files reviewed, 1 unresolved discussion (waiting on @coryan and @dbolduc)

___
*[`google/cloud/bigquery/v2/minimal/internal/dataset_client.h` line 56 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11414#gh-1179479166:-NU2zbOiDQk8EZ-pqP73:b-7bxnpl) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/89e33a8560cb98c7608784d7d3fde028de839a65/google/cloud/bigquery/v2/minimal/internal/dataset_client.h#L56)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

BTW, by convention the first paragraph in a Doxygen comment is the 'brief' description.  So it tends to be short.
</blockquote></details>

Done.

Updated comments for both apis based on the above,
___
*[`google/cloud/bigquery/v2/minimal/internal/dataset_rest_connection_impl.h` line 49 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11414#gh-1179482326:-NU2zn4h4e4sHX7dVk0g:b-do0t4y) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/89e33a8560cb98c7608784d7d3fde028de839a65/google/cloud/bigquery/v2/minimal/internal/dataset_rest_connection_impl.h#L49)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think @dbolduc had opinions as to whether these helpers actually help.
</blockquote></details>

Ack, I'll wait for Darren's comment here before submitting this.

File: google/cloud/bigquery/v2/minimal/internal/dataset_rest_connection_impl.h (hunk 2 above)
Comment (dbolduc) [reply google/cloud/bigquery/v2/minimal/internal/dataset_rest_connection_impl.h:49]: My opinion is they don't help or hurt. Feel free to use them or not.

It was a non-valuable cleanup I got distracted by. See #9168 if you are interested in the lore.

**Review by jsrinnn — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11414)* status: 0 of 13 files reviewed, 1 unresolved discussion (waiting on @coryan and @dbolduc)

___
*[`google/cloud/bigquery/v2/minimal/internal/dataset_rest_connection_impl.h` line 49 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11414#gh-1179482326:-NU34WiGE86163oEDiva:b-r4q2k1) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/89e33a8560cb98c7608784d7d3fde028de839a65/google/cloud/bigquery/v2/minimal/internal/dataset_rest_connection_impl.h#L49)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

My opinion is they don't help or hurt. Feel free to use them or not.

It was a non-valuable cleanup I got distracted by. See #9168 if you are interested in the lore.
</blockquote></details>

Ack.

I'll keep this mianly for consistency with the job apis.

I'll take a look at the link above as well. Thanks!

**Pull request merged**